Instrument the web ui for tracing

twunderlich-grapl opened the pull request:
<!-- Thank you for your contribution to Grapl! Please take some time
to fill out this short template to help us review and understand your
PR. -->

### Which issue does this PR correspond to?
https://github.com/grapl-security/issue-tracker/issues/862
https://github.com/grapl-security/issue-tracker/issues/736
<!-- Please provide a link to the GitHub Issue this pull request is
meant to address. -->

### What changes does this PR make to Grapl? Why?
This instruments the grapl-config init_grapl_env to set up tracing, and in the web-ui sets up actix auto instrumentation.


<!-- Please describe the functional impact of these changes and the
rationale behind them. This will help us understand your approach to
solving the problem. -->

### How were these changes tested?
make up
Go to localhost:1234 in your browser and click on a link
Go to http://localhost:16686/
Find web-ui on the left and search for it.
There should be traces. I'd suggest looking at the web-ui traces, of which there are two types (one using Zipkin and one using Jaeger format)
<!-- Please describe how these changes were tested. There should be
sufficient detail that reviewers can replicate your testing
procedure. If the procedure is a manual one that's OK, your
description will help us work with you to get it automated. -->

Commit:
Instrument the web ui for tracing

We're using the simple tracing instead of the more effecient batch
tracing here, because the few attempts I tried at enabling all ran into
issues where the dependencies required tokio 1.x. I plan to revisit this
after Jaeger is up and running locally.
There are plans to enable using opentelemetry with the rust tracing crate,
however again this will be worked on afterwards

wimax-grapl commented on src/rust/grapl-web-ui/src/main.rs:
```diff
@@ -57,5 +73,8 @@ async fn main() -> Result<(), GraplUiError> {
     .run()
     .await?;
 
+    // sending remaining spans. Do we need this?
+    global::shutdown_tracer_provider();
+
```
Looks like the following might do the trick instead of install_simple
https://github.com/open-telemetry/opentelemetry-rust/blob/main/examples/actix-http-tracing/src/main.rs#L33

twunderlich-grapl replied:
So the problem is that this is for a significantly earlier version of the library prior to install_simple and install_batch replacing the install method. To fix this I see a few options:
1. Downgrade the library until it works (not a huge fan since we'll want to upgrade it at some point)
2. Use an alternate runtime like async-std
3. Leave it using the simple exporter, which will work for now, but will likely need to be replaced after client 2 or 3.

I'm going to experiment with option 2. I've also found an example of someone creating a library (https://github.com/rthomas/rust-tonic-jaeger-example), with recent versions which I'm going to try to do since that's the endgoal for rust

Commit:
WIP, but proof that logger can be submitted as a layer

Tracing builds without errors but doesn't work

Commit:
Set telemetry endpoints via env vars for rust tasks

Commit:
Since we already use the init_grapl_env in our services, which sets
the logger, we should add the tracing config here as well. We're using
the simple install for the moment, but this will be switched to the
batch install later although maybe not using the tokio runtime since
the web-ui isn't on a recent enough version.

This removes the manual configuraion of open-telemetry in web-ui
since we're now using the init_grapl_env to do this

Remove web ui manual config since we're doing this in a library instead

Commit:
Remove unused dependency

Commit:
Merge branch 'main' into instrument-web-ui

inickles-grapl commented:
I plan on updating to actix-web 4 now that's been out and stable for a little bit. Sounds like that'd make some of this easier, like not having to mix tokio runtimes, which is something I'm really looking forward to removing from this.

Commit:
Merge branch 'main' into instrument-web-ui

 Conflicts:
	src/rust/grapl-web-ui/Cargo.toml
	src/rust/grapl-web-ui/src/main.rs

Commit:
Upgrade actix-open-telemetry and dependency conflicts

Now that we've upgraded actix to 4, we need to upgrade the open-telemetry
middleware so it works. In doing so, we hit a dependency conflict
from zstd, which required a version bump.

Commit:
Commit the rest of the zstd upgrades

Commit:
Instrument awc calls from web-ui

This adds a new span to the trace for the awc call. Essentially, we're
now capturing client information in this span.

Commit:
formatting

twunderlich-grapl commented on src/rust/grapl-web-ui/src/services/mod.rs:
```diff
@@ -36,11 +37,15 @@ pub(self) async fn fwd_request_to_backend_service(
-    let mut res = forwarded_req.send_stream(payload).await.map_err(|error| {
-        tracing::error!(%error);
+    let mut res = forwarded_req
+        .trace_request()
```
This adds a span for the client-side request. This can be helpful when trying to figure out why a request isn't working or reaching a service. This becomes waaaay more useful if we ever need to request a service that we don't control

Commit:
Rust services communicate with Jaeger over agent endpoints

The jaeger opentelemetry crate with tokio only supports agent
communication, not collector communication. This is why the other rust
services weren't working.
Also we apparently need to define both the port and the host for the
env vars to be picked up.

Commit:
Switch to batching traces

Batching is more performant than sending every trace on drop. Next
we'll need to add sampling

Commit:
Set the web-ui module name explicitly

Because of the mismatch in names between the nomad/consul service name
and the rust name, traces were coming in under two different service
names. This fixes that

inickles-grapl commented on src/rust/grapl-web-ui/src/main.rs:
```diff
@@ -10,18 +10,25 @@ use actix_web::{
 enum GraplUiError {
     #[error(transparent)]
     Config(#[from] config::ConfigError),
     #[error("IO error")]
     Io(#[from] std::io::Error),
+    #[error("Trace Error")]
+    Trace(#[from] TraceError),
```
@colin-grapl I have fixes this and other errors in this crate, I'll get this in a later PR.

inickles-grapl commented on src/rust/grapl-web-ui/src/services/mod.rs:
```diff
@@ -36,11 +37,15 @@ pub(self) async fn fwd_request_to_backend_service(
+        .send_stream(payload)
+        .await
+        .map_err(|error| {
+            tracing::error!(%error);
```
Similar to other comment about error cleanup.

inickles-grapl commented on src/rust/grapl-config/src/lib.rs:
```diff
@@ -93,13 +100,28 @@ pub fn region() -> Region {
+    let tracer = opentelemetry_jaeger::new_pipeline()
+        .with_service_name(service_name)
+        .install_batch(opentelemetry::runtime::Tokio)
+        .unwrap();
```
I'll make a note to catch this in error handling cleanup.

Commit:
Improve error handling

This makes it easier to debug when a program panics, using our standard
rust patterns

Commit:
Revert "Improve error handling"

This reverts commit 3099551f08362b80d6f22a6792a8c09eda48261e
because Ian has these changes in a different PR
